src/test/java/com/squareup/javapoet/TestFiler.java

```diff
@@ -30,20 +31,28 @@
 import javax.tools.FileObject;
 import javax.tools.JavaFileManager;
 import javax.tools.JavaFileObject;
-
-import static java.nio.charset.StandardCharsets.UTF_8;
+import javax.tools.SimpleJavaFileObject;
 
 final class TestFiler implements Filer {
+  static class Source extends SimpleJavaFileObject {
+    private final OutputStream stream;
+    protected Source(URI uri, OutputStream stream) {
+      super(uri, Kind.SOURCE);
+      this.stream = stream;
+    }
+    @Override public OutputStream openOutputStream() {
```

Comment:
I move the directory, file and stream creation into `Source`. Feels the 1% better.